File: python/gameOfLifeV1.py
import sys

class GameOfLifeV1:
    board = None
    rows = None
    cols = None

    def __init__(self, rows, cols, values):
        # Initialize a matrix with all values set to False
        self.rows = int(rows)
        self.cols = int(cols)

        # Pad hex string to 16 characters (64 bits) by adding zeros to the RIGHT
        hex_padded = values.ljust(16, '0')
        _values = int(hex_padded, 16)
        binary_str = format(_values, '064b')

        # Create board from binary string (reverse bits within each byte)
        self.board = []
        for i in range(8):
            byte_bits = binary_str[i*8:(i+1)*8]
            row = [bit == '1' for bit in byte_bits]
            self.board.append(row)


    def print(self):
        bits_by_row = []
        for row in self.board:
            row_bits = "".join(str(int(b)) for b in row)
            bits_by_row.append(row_bits)
        bits = "".join(bits_by_row)
        hex_value = f"{int(bits, 2):016X}"
        print(hex_value)

    def iterate(self, next_board):
        for i in range(self.rows):
            for j in range(self.cols):
                next_board[i][j] = False

        for i in range(1, len(self.board)-1):
            for j in range(1, len(self.board[i])-1):
                num = self.calculateNeighbors(i, j)
                if num == 3 or (num == 2 and self.board[i][j]):
                    next_board[i][j] = True
                else:
                    next_board[i][j] = False

    def calculateNeighbors(self, i, j):
        return (self.board[i-1][j-1] + self.board[i-1][j] + self.board[i-1][j+1] 
              + self.board[i][j-1]                        + self.board[i][j+1] 
              + self.board[i+1][j-1] + self.board[i+1][j] + self.board[i+1][j+1])
    

if __name__ == "__main__":
    if len(sys.argv) != 3:
        print("Usage: python main.py <param1> <param2>")
        sys.exit(1)
    
    executions = int(sys.argv[1])
    initialvalue = sys.argv[2]
    x = 8
    y = 8

    game = GameOfLifeV1(x, y, initialvalue)
    next_board = [[False for _ in range(y)] for _ in range(x)]

    for i in range(executions):
        game.iterate(next_board)
        game.board, next_board = next_board, game.board  # Swap
    game.print()

File: java/GameOfLifeV1.java
public class GameOfLifeV1 {
    private boolean[][] board;
    private int rows;
    private int cols;

    public GameOfLifeV1(int rows, int cols, String values) {
        this.rows = rows;
        this.cols = cols;

        // Pad hex string to 16 characters (64 bits) by adding zeros to the RIGHT
        String hexPadded = String.format("%-16s", values).replace(' ', '0');

        // Convert hex to binary string (64 bits)
        long longValue = Long.parseUnsignedLong(hexPadded, 16);
        String binaryStr = String.format("%64s", Long.toBinaryString(longValue)).replace(' ', '0');

        // Create board from binary string
        this.board = new boolean[8][8];
        for (int i = 0; i < 8; i++) {
            for (int j = 0; j < 8; j++) {
                int index = i * 8 + j;
                this.board[i][j] = binaryStr.charAt(index) == '1';
            }
        }
    }

    public void print() {
        StringBuilder bits = new StringBuilder();
        for (int i = 0; i < this.rows; i++) {
            for (int j = 0; j < this.cols; j++) {
                bits.append(this.board[i][j] ? '1' : '0');
            }
        }

        long longValue = Long.parseUnsignedLong(bits.toString(), 2);
        String hexValue = String.format("%016X", longValue);
        System.out.println(hexValue);
    }

    public void iterate(boolean[][] nextBoard) {
        // Limpiar next_board
        for (int i = 0; i < this.rows; i++) {
            for (int j = 0; j < this.cols; j++) {
                nextBoard[i][j] = false;
            }
        }

        for (int i = 1; i < this.board.length - 1; i++) {
            for (int j = 1; j < this.board[i].length - 1; j++) {
                int num = calculateNeighbors(i, j);
                if (num == 3 || (num == 2 && this.board[i][j])) {
                    nextBoard[i][j] = true;
                } else {
                    nextBoard[i][j] = false;
                }
            }
        }
    }

    private int calculateNeighbors(int i, int j) {
        int count = 0;
        count += this.board[i-1][j-1] ? 1 : 0;
        count += this.board[i-1][j]   ? 1 : 0;
        count += this.board[i-1][j+1] ? 1 : 0;
        count += this.board[i][j-1]   ? 1 : 0;
        count += this.board[i][j+1]   ? 1 : 0;
        count += this.board[i+1][j-1] ? 1 : 0;
        count += this.board[i+1][j]   ? 1 : 0;
        count += this.board[i+1][j+1] ? 1 : 0;
        return count;
    }

    public void setBoard(boolean[][] newBoard) {
        this.board = newBoard;
    }

    public boolean[][] getBoard() {
        return this.board;
    }

    public static void main(String[] args) {
        if (args.length != 2) {
            System.err.println("Usage: java GameOfLifeV1 <param1> <param2>");
            System.exit(1);
        }

        int executions = Integer.parseInt(args[0]);
        String initialValue = args[1];
        int x = 8;
        int y = 8;

        GameOfLifeV1 game = new GameOfLifeV1(x, y, initialValue);
        boolean[][] nextBoard = new boolean[x][y];

        for (int i = 0; i < executions; i++) {
            game.iterate(nextBoard);
            // Swap
            boolean[][] temp = game.getBoard();
            game.setBoard(nextBoard);
            nextBoard = temp;
        }

        game.print();
    }
}
